pghoard/common.py

```diff
@@ -100,6 +104,59 @@ class BaseBackupMode(StrEnum):
     pipe = "pipe"
 
 
+class ProgressData(BaseModel):
+    current_progress: float = 0
+    last_updated_time: float = 0
+
+    @property
+    def age(self) -> float:
+        return time.time() - self.last_updated_time
+
+    def update(self, current_progress: float) -> None:
+        self.current_progress = current_progress
+        self.last_updated_time = time.time()
+
+
+class PersistedProgress(BaseModel):
+    progress: Dict[str, ProgressData] = Field(default_factory=dict)
+    _lock: threading.Lock = threading.Lock()
+
+    @classmethod
+    def read(cls, metrics: Metrics) -> "PersistedProgress":
+        if os.path.exists(PROGRESS_FILE):
+            with open(PROGRESS_FILE, "r") as file:
+                try:
+                    return cls.parse_raw(file.read())
+                except Exception as ex:  # pylint: disable=broad-except
+                    LOG.exception("Failed to read persisted progress file: %r", ex)
+                    metrics.unexpected_exception(ex, where="read_persisted_progress")
+        return cls()
+
+    def write(self, metrics: Metrics):
+        with self._lock:
+            try:
+                with tempfile.NamedTemporaryFile("w", delete=False, dir=os.path.dirname(PROGRESS_FILE)) as temp_file:
```

Comment:
Would be great if this atomic file write is extracted into a separate function, this functionality is used in a couple of places already. (Search for os.rename)
Also cleanup of temp file is missing in case of exception.

Reply:
yes, I will move it to a separate function, but the atomic file write is used differently in other places... Also, remove the temp file in case of exception.